packages/commonwealth/client/scripts/views/pages/notification_settings/index.tsx

```diff
@@ -56,6 +60,22 @@ const NotificationSettingsPage = () => {
     relevantSubscribedCommunities,
   } = useNotificationSettings();
 
+  const { mutateAsync: registerClientRegistrationToken } =
+    trpc.subscription.registerClientRegistrationToken.useMutation();
+
+  const handlePushNotificationSubscription = useCallback(() => {
```

Comment:
This doesn't need to be a callback, it could be a regular async function.

Reply:
Unfortunately, it can't... you can't use an async function in a React onClick event handler without it losing the error.  This is a common mistake and why I added the eslint promise rules. The function will be called but there's no defined behavior on a promise that's called without a .then() .catch() or an await.

Reply:
.. assuming I understood what you meant.

Reply:
Strange. It works for me. You can try a mini example to reproduce the behaviour:
```
    <button onClick={() => Promise.reject('promise failed on click')}>
      Hello world
    </button>
```

If you click this button, it is what you would expect from a failed promise
<img width="1443" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/14794654/aebf735d-cea3-44c8-b6b7-50a5c0cdddb3">

Reply:
I think we are talking about two different things here. There is the way javascript works, and then there is the UI snack. But the code that you have currently will also not show a snack when it fails because the logic is not coded in.

But regardless of the snack, I am suggesting you remove the useCallback along with the doAsync().catch(console.error) and just use the doAsync directly. This is because this will have the same functionality but with less code.

Reply:
This is why I like to break out these PRs so that we don't get stuck on issues like this.  :)

I believe eslint will block use of:

<button onClick={reject}>

and we should because the main reason is because there is no way to know if you're properly handling errors in the above code.

IMO, after null pointer errors and off by 1 errors, incorrect promise use is the biggest bug facing frontend applications.

You're right in the above code that it will work but what if someone does this:

<button onClick={doAsyncWhichFails}>click me</button>

There's no way to catch that and realize that it has failed and no way to block it in the build so the dangerous code will make it into production.

Also, without eslint rules blocking this, even if you KNOW about this behavior, you still might accidentally pass a async function to a sync event handler. 

Honestly, I think promises and async handling is handled incorrectly so often that it should be considered an anti-pattern. It's basically constantly running with scissors.

The code I implemented is far more safe because:

- It's using useCallback properly so I'm not contributing to the react re-render storm problem.
- I can just slap in the errorLogger at any time without rewriting code.

The other thing I feel pretty strongly about is useCallback and useMemo just being necessary evils that you should try to use as much as possible but with React 19 we won't need them anymore.

Reply:
> I believe eslint will block use of:

Ah my bad, you are correct here. To satisfy the eslint rule this it could be:

```
<button onClick={() => { Promise.reject('promise failed on click').catch(console.error) } }>
      Hello world
    </button>
```

In your case it would be:
```
const handlePushNotificationSubscription = async () => {
const permission = await Notification.requestPermission();
      if (permission === 'granted') {
        console.log('Notification permission granted.');
        const token = await getFirebaseMessagingToken();
        await registerClientRegistrationToken.mutateAsync({
          id: 'none',
          token,
        });
      }
};
...

onClick={handlePushNotificationSubscription.catch(console.error)}
```

But the crux of my argument is around the use of useCallback.

> I'm not contributing to the react re-render storm problem

Where exactly is the re-render storm here, given that nothing updates state inside this callback? No matter how many times this code is ran, it will not cause any other components to re-render.

Reply:
If you pass variables from React components A-> B -> C then a non-memoized variable in A could cause C to re-render.  If you have a tree of objects these can cause serious "render storms" and your app will start to become very very slow.

To avoid this I always memoize because the downside can be REALLY bad and you don't really see it coming until it's too late. Then you have to refactor a massive amount of code.

Reply:
Also. I think you mean:

```typescript

onClick={() => handlePushNotificationSubscription.catch(console.error)}
.catch(console.error)
```

Reply:
Yeah ok it is probably fine, but we should discuss it next Product office hours, I will drop in. I think it is best to have a standard for it going forward so that the code looks more unified (Either always use useCallback, or only use it when necessary).